src/Spanner/Session/CacheSessionPool.php

```diff
@@ -0,0 +1,548 @@
+     * @param array $data
+     * @param bool $hasSession
+     * @return array
+     */
+    private function buildToCreateList(array $data, $hasSession)
+    {
+        $number = 0;
+        $toCreate = [];
+        $time = $this->time();
+        $count = $this->getSessionCount($data);
+
+        if ($count < $this->config['minSessions']) {
+            $number = $this->config['minSessions'] - $count;
+        } elseif (!$hasSession && !$data['queue'] && $count < $this->config['maxSessions']) {
+            $number++;
+        }
+
+        for ($i = 0; $i < $number; $i++) {
+            $toCreate[uniqid($time . '_')] = $time;
+        }
+
+        return $toCreate;
+    }
+
+    /**
+     * Purge any items in the to create queue that have been inactive for 20
+     * minutes or more.
+     *
+     * @param array $data
+     */
+    private function purgeOrphanedToCreateItems(array &$data)
+    {
+        foreach ($data['toCreate'] as $key => $timestamp) {
+            if ($timestamp + 1200 < $this->time()) {
+                unset($data['toCreate'][$key]);
+            }
+        }
+    }
+
+    /**
+     * Purge any in use sessions that have been inactive for 20 minutes or more.
+     *
+     * @param array $data
+     */
+    private function purgeOrphanedInUseSessions(array &$data)
+    {
+        foreach ($data['inUse'] as $key => $session) {
+            if ($session['lastActive'] + 1200 < $this->time()) {
+                unset($data['inUse'][$key]);
+            }
+        }
+    }
+
+    /**
+     * Initialize the session data.
+     *
+     * @return array
+     */
+    private function initialize()
+    {
+        return [
+            'queue' => [],
+            'inUse' => [],
+            'toCreate' => []
+        ];
+    }
+
+    /**
+     * Returns the total count of sessions in queue, use, and in the process of
+     * being created.
+     *
+     * @param array $data
+     * @return int
+     */
+    private function getSessionCount(array $data)
+    {
+        $count = 0;
+
+        foreach ($data as $sessionType) {
+            $count += count($sessionType);
+        }
+
+        return $count;
+    }
+
+    /**
+     * Gets the next session in the queue, clearing out which are expired.
+     *
+     * @param array $data
+     * @return array|null
+     */
+    private function getSession(array &$data)
+    {
+        $session = array_shift($data['queue']);
+
+        if ($session) {
+            if ($session['expiration'] - 60 < $this->time()) {
+                return $this->getSession($data);
+            }
+
+            $data['inUse'][$session['name']] = $session + [
+                'lastActive' => $this->time()
+            ];
+        }
+
+        return $session;
+    }
+
+    /**
+     * Creates sessions up to the count provided.
+     *
+     * @param int $count
+     * @return array
+     */
+    private function createSessions($count)
+    {
+        $sessions = [];
+
+        for ($i = 0; $i < $count; $i++) {
```

Comment:
Does it make sense to limit the maximum number of sessions it creates at one time? For example, even if there is a room for 100 sessions to create in the pool, can we just create 5 sessions so that the client will very likely be able to get one of them?